nautobot_floor_plan/forms.py

```diff
@@ -217,34 +228,105 @@ def clean(self):
-
-            # Clear existing custom ranges
-            models.FloorPlanCustomAxisLabel.objects.filter(floor_plan=instance).delete()
-
-            # Save X axis custom ranges
-            if x_ranges:
-                self.create_custom_axis_labels(x_ranges, instance, axis="X")
-            # Save Y axis custom ranges
-            if y_ranges:
-                self.create_custom_axis_labels(y_ranges, instance, axis="Y")
+    def _post_clean(self):
+        """Process formset data after Django's validation."""
+        super()._post_clean()
+
+        def process_ranges(formset, axis):
+            """Process formset data into a list of range dictionaries."""
+            ranges_data = []
+
+            for form in formset.forms:
+                if not form.is_valid():
+                    continue  # Skip invalid forms
+
+                delete = form.cleaned_data.get("DELETE", False)
+                if delete:
+                    continue  # Skip deleted forms
+
+                start = form.cleaned_data.get("start", "")
+                end = form.cleaned_data.get("end", "")
+
+                if start and end:
+                    range_data = {
+                        "start": start,
+                        "end": end,
+                        "step": form.cleaned_data.get("step", 1),
+                        "label_type": form.cleaned_data.get("label_type"),
+                        "increment_letter": form.cleaned_data.get("increment_letter", False),
+                        "form_index": formset.forms.index(form),
+                    }
+                    ranges_data.append(range_data)
+
+            if ranges_data:
+                try:
+                    validated_ranges = self._validate_custom_ranges(ranges_data)
+                    if validated_ranges:
+                        self.cleaned_data[f"{axis}_custom_ranges"] = json.dumps(validated_ranges)
+                    return validated_ranges
+                except forms.ValidationError as e:
+                    form = formset.forms[ranges_data[0]["form_index"]]
+                    error_field = "start" if "start" in str(e).lower() else "end"
+                    form.add_error(error_field, e)
+                    return []
+
+            return ranges_data
+
+        # Process X and Y ranges
+        process_ranges(self.x_ranges, "X")
+        process_ranges(self.y_ranges, "Y")
+
+    def is_valid(self):
+        """Custom is_valid to ensure both form and formsets are valid."""
+        # Check main form validity
+        form_valid = super().is_valid()
+
+        # Check formset validity
+        x_ranges_valid = self.x_ranges.is_valid()
+        y_ranges_valid = self.y_ranges.is_valid()
+
+        # Only return True if everything is valid
+        is_valid = all(
+            [
+                form_valid,
+                x_ranges_valid,
+                y_ranges_valid,
+                not any(form.errors for form in self.x_ranges.forms),
```

Comment:
What would be a case where, for example, `x_ranges_valid` is True but one of its subforms has errors?